compiler/rustc_trait_selection/src/solve/search_graph/mod.rs

```diff
@@ -221,27 +232,37 @@ impl<'tcx> SearchGraph<'tcx> {
         canonical_goal: CanonicalGoal<'tcx>,
         mut loop_body: impl FnMut(&mut Self) -> QueryResult<'tcx>,
     ) -> QueryResult<'tcx> {
+        if let Some(result) = tcx.new_solver_evaluation_cache.get(&canonical_goal, tcx) {
+            return result;
+        }
+
         match self.try_push_stack(tcx, canonical_goal) {
             Ok(()) => {}
             // Our goal is already on the stack, eager return.
             Err(response) => return response,
         }
 
-        self.repeat_while_none(
-            |this| {
-                let result = this.deal_with_overflow(tcx, canonical_goal);
-                let stack_elem = this.stack.pop().unwrap();
-                this.try_move_finished_goal_to_global_cache(tcx, stack_elem);
-                result
-            },
-            |this| {
-                let result = loop_body(this);
-                if this.try_finalize_goal(tcx, canonical_goal, result) {
-                    Some(result)
-                } else {
-                    None
-                }
-            },
-        )
+        // This is for global caching, so we properly track query dependencies.
+        // Everything that affects the `Result` should be performed within this
+        // `with_anon_task` closure.
+        //
+        // We only cache the cycle root (why?)
+        let (result, dep_node) = tcx.dep_graph.with_anon_task(tcx, DepKind::TraitSelect, || {
+            self.repeat_while_none(
+                |this| {
+                    let result = this.deal_with_overflow(tcx, canonical_goal);
+                    let _ = this.stack.pop().unwrap();
+                    result
+                },
+                |this| {
+                    let result = loop_body(this);
+                    this.try_finalize_goal(canonical_goal, result).then(|| result)
+                },
+            )
+        });
+
+        self.try_move_finished_goal_to_global_cache(tcx, canonical_goal, dep_node);
+
+        result
     }
```

Comment:
In terms of code organisation, it may be interesting to have a single function that does all the global cache manipulation (looking at the cached goals, calling the dep-graph and containing the code in `try_move_finished_goal_to_global_cache`), and eventually generic on the closure that you pass to `with_anon_task`.

```
    fn with_new_goal(
        &mut self,
        tcx: TyCtxt<'tcx>,
        canonical_goal: CanonicalGoal<'tcx>,
        evaluate: impl FnMut(&mut Self) -> QueryResult<'tcx>,
    ) -> QueryResult<'tcx> {
        if let Some(result) = tcx.new_solver_evaluation_cache.get(&canonical_goal, tcx) {
            return result;
        }

        match self.try_push_stack(tcx, canonical_goal) { ... }

        let (result, dep_node) = tcx.dep_graph.with_anon_task(tcx, DepKind::TraitSelect, evaluate);

        /* inline try_move_finished_goal_to_global_cache */

        result
    }
```

Reply:
Ended up inlining `try_move_finished_goal_to_global_cache` here. Also attempted to distill that comment about why we can't cache non-root goals, but not sure if I totally wrap my head around it despite a thorough explanation :sweat_smile: